keras/src/dtype_policies/dtype_policy_map.py

```diff
@@ -98,35 +98,64 @@ def quantization_mode(self):
-        there are multiple matches for the regex, an `ValueError` will be
-        raised. Returns `self.default_policy` if there isn't any match found.
+        This method first attempts an exact key match. If no exact match is
+        found, it treats the keys stored in the map as regular expression
+        patterns and searches for a match against the input `key`.
+
+        A regex match is only considered valid if it meets two conditions:
+        - It must match exactly one pattern in the policy map.
+        - The policy associated with that pattern must not be for quantization.
 
         Args:
-            key: String key to query a `DTypePolicy`.
+            key: The string key to query for a `DTypePolicy`.
 
         Returns:
-            Corresponding `DTypePolicy` based on the query.
+            The corresponding `DTypePolicy`. If no valid match is found
+            (either exact or regex), this method returns `self.default_policy`.
+
+        Raises:
+            ValueError: If the `key` matches more than one regex pattern.
         """
+        # 1. Check for an exact match.
         if key in self._policy_map:
             return self._policy_map[key]
 
-        matching_keys = []
-        for k in self._policy_map:
-            if re.search(k, key):
-                matching_keys.append(k)
+        # 2. If no exact match is found, treat the keys in the map as regular
+        # expression patterns and search for a match against the input `key`.
+        matching_keys = [
+            pattern for pattern in self._policy_map if re.search(pattern, key)
+        ]
+
+        # 3. Handle cases based on the number of matches found.
         if len(matching_keys) > 1:
             raise ValueError(
                 f"Path '{key}' matches multiple dtype policy "
                 f"specification keys: {matching_keys}. Please make "
                 "sure each path only matches at most "
                 "one dtype policy specification key in the DTypePolicyMap."
             )
-        elif len(matching_keys) == 1:
-            return self._policy_map[matching_keys[0]]
+
+        if len(matching_keys) == 1:
+            policy = self._policy_map[matching_keys[0]]
+
+            # A single regex match is only considered valid if the resulting
+            # policy is non-quantized.
+            # This logic specifically handles how KerasHub serializes quantized
+            # models. In that workflow, every quantized layer is stored in the
+            # policy map with its exact path, ensuring it's found by the direct
+            # lookup at the start of this method.
+            #
+            #  For example, it stops a general key like "attention/query" which
+            # represents a quantization-compatible layer from incorrectly
+            # matching a different, incompatible layer like
+            # "attention/query_norm"
+            if not policy.quantization_mode:
```

Comment:
I'm not sure I'm following the exact bug here, but this seems like we are losing a lot of flexibility with the API here. The hope was this could also be used as a more granular dtype setting API for non quantized cases, say running at different precisions for different layers.

```
dtype_policy_map = DTypePolicyMap()
dtype_policy_map["layer/dense_0"] = DTypePolicy("float32")
dtype_policy_map["layer/dense_1"] = DTypePolicy("bfloat16")
```

Won't we lose that ability with this change? Is there another way we can fix whatever the issue is here?

Reply:
As discussed offline, I've replaced the logic to instead use a more robust regex matching logic instead of incorporating special handling for quantized dtypes which makes this behavior easier to understand and predictable.